libkmod/libkmod-config.c

```diff
@@ -151,15 +151,6 @@ static int kmod_config_add_command(struct kmod_config *config, const char *modna
 	return 0;
 }
 
-static void kmod_config_free_command(struct kmod_config *config, struct kmod_list *l,
-				     struct kmod_list **list)
-{
-	struct kmod_command *cmd = l->data;
-
-	free(cmd);
-	*list = kmod_list_remove(l);
-}
-
```

Comment:
What is "inconsistent" about the wrappers? I have removed the ones which also call `kmod_list_remove`. I don't see the point in simplifying the ones which currently just call free. If your point is that we can have a `kmod_list_release` which implicitly calls `free`, this dangerously implies that every list could just be freed, which is not correct, i.e. `kmod_module_unref`, and limits further usages of `kmod_list_release` to data which can be just freed.

Reply:
> we can have a kmod_list_release which implicitly calls free, this dangerously implies that every list could just be freed

This can trivially be addressed with by using more indicative name (kmod_list_release_free, kmod_list_free, etc) and adding 2-3 words of documentation what the function does.

Reply:
Doing things trivially doesn't mean that things are done correctly. In this case, the API is extended in a universally included header for non-universal usage (it's either `free` or `kmod_module_unref`). Please read the current kmod-list code. No single function cares about the content of `list->data`. Not even `kmod_list_remove_data` which just compares pointers. The kmod-list implementation is as agnostic as it can be towards data included in the list. It even goes to the extend that the list information is `malloc`ed apart from data. How would a `kmod_list_release` function fit into this which just calls `free`? And yes, it could be documented, but this still doesn't fit into the API at all. List and array simply don't care what they keep inside them. From a different point of view: Are we missing a `read_and_close` system call with its own manual page? I don't think so.

Reply:
It's fine to have different proposal (and thus mental model) how things would be better.

Replies like this come as "my way or the highway" and somewhat patronising which doesn't seem very constructive IMHO.

Hyperbole like the `read_and_close` are also not particularly helpful.

Reply:
On the technical side - both proposals suffer from somewhat opposing issues IMHO: 

One proposal is aimed to address the common case both in terms of de-duplication and passing the correct free function - 15 out of the 16 cases, if I haven't failed at counting. It also benefits from the (admittedly rather loose) C type check system by making it a function.

The other does more or less the complete opposite - all cases covered, yet it's easier to use the wrong free function, plus type checking practically is non-existent in macros.

There are other approaches which will resolve the above although they have their own caveats.

For example:
Having a distinct list types with designated `_release` API + others as needed. Implementation wise it's a typecast or a simple wrapper, if we'd want to add some validation. One could also extend the official libkmod API to use it and enforce some basic validation.

In other words:
```
$ cat libkmod.h
...
int kmod_module_unref_list(struct kmod_list *list); // OLD
int kmod_module_unref_list(struct kmod_module_list *list); // NEW
...

$ cat libkmod-internal.h

// no validation
typedef struct kmod_modulr_list struct kmod_list;
void kmod_module_list_release(struct kmod_module_list *list);

// validation
struct kmod_module_list {
    int magic; // aka 0xDEADBEEF or alike
    struct kmod_list;
}
...
```

The pros:
 - clarity for libkmod users - oops I passed a module symbols kmod_list where the API was expecting a modules kmod_list
 - some type-safety and API which is harder to misuse

Cons:
 - bit of extra code, although manageable
 - few CPU cycles of overhead though, if we choose extra validation ...
 - how to make the libkmod.h changes while preserving API _and_ having the extra validation

Reply:
> The other does more or less the complete opposite - all cases covered, yet it's easier to use the wrong free function, plus type checking practically is non-existent in macros.

What do you mean with type checking here? If you are talking about what the C programming language implies, it has full type checking, because the macro is extended by preprocessor. Replace `free` with `exit` and see that the compiler tells you exactly that you are using the wrong function. This is how I ended up with a macro instead of an inline function with a function pointer (they are just horrible and such an easy step for an attacker to get control).

> For example: Having a distinct list types with designated `_release` API + others as needed.

That would be basically the `SLIST` family of macros in `queue.h`. I'm not sure why they are not used. In fact, I don't try to defend kmod-list here. Calling depmod leads to tons of memory allocations due to these lists, which allocate every single node AND the data. Neither did I have a good idea how to overcome this without breaking the API nor did I see the need yet, since there is not much of a performance issue right now to tackle such a huge part of the API.

> * clarity for libkmod users - oops I passed a module symbols kmod_list where the API was expecting a modules kmod_list
> * some type-safety and API which is harder to misuse

Yes, fully agree here.

> Cons:
> 
> * how to make the libkmod.h changes while preserving API _and_ having the extra validation

And here as well.

And coming back to the subject of this PR: Adjusting the whole API for releasing memory on error paths is way out of proportion. I decided to stay close to "the spirit of kmod-list" as possible.

If my explanations how I came to my conclusions before opening the PR, my arguments and examples are perceived as patronising, then it's definitely not my intention. I try to highlight how I ended up with this code and if there are no arguments next to "we could also do it this way," then I'm hardly convinced to give more thoughts to such an alternative.

Reply:
The following illustrates both issues:

```c
struct foo {
  void *list; // not a kmod_list
  int baz;
} f = ..., *foo = &f;

kmod_list_release(foo->list, foobar_ref); // typo'd unref
```

I'm 110% on fixing the leaks and adding proper abstraction is indeed out of topic. Since we don't have an agreement on the intermediate bits - let's just fix the leaks and worry about the rest at later point.